### PR TITLE
Use PHP-Scoper to prefix production dependencies in a third-party directory

### PR DESCRIPTION
**Pull request by felixarntz**
## Summary

Addresses issue #612

~~Note that after this is merged, we need to update the contributor guidelines on getting started to include `composer global require humbug/php-scoper`~~

## Relevant technical choices

* Use [PHP-Scoper](https://packagist.org/packages/humbug/php-scoper) because it's more reliable and better maintained than alternatives such as [Mozart](https://packagist.org/packages/coenjacobs/mozart).
* Whitelist vendor content needed for production in `scoper.inc.php` configuration file.
* Generate individual classmap autoloaders for both Site Kit's own files and for all prefixed dependencies.
* Hack a bit regarding the non-class files that dependencies include. We can simply use the `autoload_files.php` file generated by the original autoloader and load them in a bit different way. This at least ensures we don't have to manually maintain that list of files.

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

File: .travis.yml [hunk 1]
```diff
@@ -51,7 +51,7 @@ before_install:
 
 before_script:
   - npm install || exit 1
-  - composer install
+  - docker run --rm -v "$PWD:/app" composer install
   - |
     if [[ "$PHP" == "1" ]] || [[ "$JS" == "1" ]] || [[ "$SNIFF" == "1" ]]; then
       npm install -g gulp-cli
```
Comment (aaemnnosttv) [inline .travis.yml]: These things should run in the `install` phase of the build rather than `before_script` I think

File: scoper.inc.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+			'Analytics',
+			'AnalyticsReporting',
+			'AdSense',
+			'Pagespeedonline',
+			'PeopleService',
+			'SiteVerification',
+			'TagManager',
+			'Webmasters',
+		)
+	)
```
Comment (aaemnnosttv) [inline scoper.inc.php]: This `array_map` seems unnecessary. Everything within is already regex-safe.

File: scoper.inc.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+return array(
+	'prefix'                     => 'Google\\Site_Kit_Dependencies',
+	'finders'                    => array(
+
+		// General dependencies, except Google API services.
+		Finder::create()
+			->files()
+			->ignoreVCS( true )
+			->notName( '/LICENSE|.*\\.md|.*\\.dist|Makefile|composer\\.json|composer\\.lock/' )
```
Comment (aaemnnosttv) [inline scoper.inc.php]:
```suggestion
			->notName( '/LICENSE|.*\\.md|.*\\.dist|Makefile|composer\\.(json|lock)/' )
```

File: scoper.inc.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+			->path( '#^psr/#' )
+			->path( '#^ralouphie/#' )
+			->path( '#^react/#' )
+			->in( 'vendor' ),
+
+		// Google API service infrastructure classes.
+		Finder::create()
+			->files()
+			->ignoreVCS( true )
+			->notName( '/LICENSE|.*\\.md|.*\\.dist|Makefile|composer\\.json|composer\\.lock/' )
```
Comment (aaemnnosttv) [inline scoper.inc.php]:
```suggestion
			->notName( '/LICENSE|.*\\.md|.*\\.dist|Makefile|composer\\.(json|lock)/' )
```

File: scoper.inc.php (hunk 2 above)
Comment (felixarntz) [reply scoper.inc.php]: I mostly put it there to be safe for future changes, but agreed, we can also do so once actually needed.

File: scoper.inc.php (hunk 4 above)
Comment (tofumatt) [reply scoper.inc.php]: Looks like this change should still be made.

**Comment by felixarntz** [comment]
@tofumatt I committed that change already, not sure why it still shows up, but it's done :)

**Pull request merged**